contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,158 @@
+#include <eosio.system/eosio.system.hpp>
+
+#include <eosio/eosio.hpp>
+#include <eosio/privileged.hpp>
+#include <eosio/crypto_bls_ext.hpp>
+
+namespace eosiosystem {
+
+   bool system_contract::is_savanna_consensus() {
+      return !_gstate5.last_finalizers.empty();
+   }
+
+   bool system_contract::is_registered_producer( const name& producer ) {
+      // Returns true if `producer` is a registered producer
+   }
+
+   bool system_contract::is_active_producer( const name& producer ) {
+      // Returns true if `producer` is an active producer (top 21)
+   }
+
+   bool set_next_finalizers_set(const std::vector<name>& producers ) {
+      // Build finalizer set
+      std::vector<finalizer_authority> finalizers;
+      for ( const auto p&: producers ) {
+         auto it = _finalizer_keys.find( p.value );
+         if( it != _finalizer_keys.end() ) { // p has registered a finalizer key
+            finalizers.emplace_back(
+               finalizer_authority{
+                  .description = "",
+                  .weight      = 1,
+                  .public_key  = it->active_key
+               }
+            );
+         }
+      }
+
+      // Abort if less than 15 top producers have registered a finalizer key
+      if( finalizers.size < 15 ) {
+         return false;
+      }
+
+      // Establish finalizer policy and call set_finalizers() host function
+      finalizer_policy fin_policy {
+         .generation = 0;
+         .threshold  = 15;
+         .finalizers = finalizers;
+      };
+      set_finalizers(std::move(fin_policy));
+
+      _gstate5.last_finalizers = finalizers;
+
+      return true;
+   }
+
+   void replace_key_in_finalizer_policy(const name& producer, const std::string& finalizer_key) {
+      // replace the key in last finalizer policy and call set_finalizers
+      // host function immediately
+   }
+
+   void delete_key_in_finalizer_policy(const name& producer, const std::string& finalizer_key) {
+      // replace the key in last finalizer policy and call set_finalizers
+      // host function immediately
+   }
+
+   void system_contract::switchtosvnn() {
+      require_auth(get_self());
+
+      check(!is_savanna_consensus(), "switchtosvnn can only be run once");
+
+      // Find top 21 block producers
+      auto idx = _producers.get_index<"prototalvote"_n>();
+      std::vector<name> top_producers;
+      top_producers.reserve(21);
+      for( auto it = idx.cbegin(); it != idx.cend() && top_producers.size() < 21 && 0 < it->total_votes && it->active(); ++it ) {
```

Comment:
We need to be consistent with the selection process used here as we do in `update_elected_producers`.